// sonarqube-scanner/src/deps.js
/*
 * sonar-scanner-npm
 * Copyright (C) SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * You can redistribute and/or modify this program under the terms of
 * the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
import { exec, spawn } from 'node:child_process';
import fs from 'node:fs';
import util from 'node:util';
import { extractArchive } from './file.js';
import { fetchJRE, serverSupportsJREProvisioning } from './java.js';
import { locateExecutableFromPath } from './process.js';
import { download as defaultDownload, fetch as defaultFetch } from './request.js';
import { downloadScannerCli, runScannerCli } from './scanner-cli.js';
import { fetchScannerEngine, runScannerEngine } from './scanner-engine.js';
const execAsync = util.promisify(exec);
/**
 * Creates the default dependencies using real implementations.
 */
function createDefaultDeps() {
    return {
        fs: {
            existsSync: fs.existsSync,
            statSync: fs.statSync,
            readFileSync: fs.readFileSync,
            readFile: fs.readFile,
            mkdirSync: fs.mkdirSync,
            createReadStream: fs.createReadStream,
            createWriteStream: fs.createWriteStream,
            remove: (filePath) => fs.promises.rm(filePath, { recursive: true, force: true }),
            writeFile: (filePath, data) => fs.promises.writeFile(filePath, data),
            exists: async (filePath) => {
                try {
                    await fs.promises.access(filePath);
                    return true;
                }
                catch {
                    return false;
                }
            },
            ensureDir: (dirPath) => fs.promises.mkdir(dirPath, { recursive: true }).then(() => { }),
        },
        process: {
            get platform() {
                return process.platform;
            },
            get arch() {
                return process.arch;
            },
            get env() {
                return process.env;
            },
            cwd: () => process.cwd(),
        },
        http: {
            fetch: defaultFetch,
            download: defaultDownload,
        },
        spawn,
        execAsync,
        scan: {
            serverSupportsJREProvisioning,
            fetchJRE,
            downloadScannerCli,
            runScannerCli,
            fetchScannerEngine,
            runScannerEngine,
            locateExecutableFromPath,
        },
        file: {
            extractArchive,
        },
    };
}
// Module-level dependency container
let deps = createDefaultDeps();
/**
 * Get the current dependency container.
 * Internal functions use this to access dependencies.
 */
export function getDeps() {
    return deps;
}
/**
 * Set/override dependencies. Used for testing.
 * Merges the provided partial dependencies with the defaults.
 *
 * @param newDeps - Partial dependencies to override
 */
export function setDeps(newDeps) {
    const defaults = createDefaultDeps();
    deps = {
        ...defaults,
        ...newDeps,
        // Deep merge nested objects
        fs: {
            ...defaults.fs,
            ...newDeps.fs,
        },
        process: {
            ...defaults.process,
            ...newDeps.process,
        },
        http: {
            ...defaults.http,
            ...newDeps.http,
        },
        scan: {
            ...defaults.scan,
            ...newDeps.scan,
        },
        file: {
            ...defaults.file,
            ...newDeps.file,
        },
    };
}
/**
 * Reset dependencies to defaults. Should be called in afterEach() in tests.
 */
export function resetDeps() {
    deps = createDefaultDeps();
}
